counter/src/main/java/org/infinispan/counter/impl/function/FunctionHelper.java

```diff
@@ -47,8 +47,22 @@ static Object compareAndSwap(EntryView.ReadWriteEntryView<?, CounterValue> entry
       return retVal;
    }
 
+   static CounterValue set(EntryView.ReadWriteEntryView<?, CounterValue> entry,
+                           CounterConfigurationMetaParam metadata, long update) {
+      CounterValue newValue;
+      if (metadata.get().type() == CounterType.BOUNDED_STRONG && update < metadata.get().lowerBound()) {
+         newValue = newCounterValue(update, CounterState.LOWER_BOUND_REACHED);
```

Comment:
We need to document the behavior with bounded counters. We can do:
* if `update` is outside the bounds, throw an exception
* if `update` is outside the bounds, set the value to the upper or lower bound.

In any case, setting the counter value to `update` is wrong. The first option seems the better one.